File: sort_algorithm/insert_sort.py
# -*- coding:utf-8 -*-


def insert_sort(arr):
    """ 插入排序算法实现 """
    
    for i in range(1, len(arr)):
        for j in range(i, 0, -1):
            if arr[j] < arr[j-1]:
                arr[j], arr[j-1] = arr[j-1], arr[j]
            else:
                break
    return arr

if __name__ == "__main__":

    print(insert_sort([54, 26, 93, 17, 77, 31, 44, 55, 20]))

File: sort_algorithm/insert_sort.cc
#include <iostream>

using namespace std;

template<typename T>
void insertSort(T arr[], int n){

    for (int i = 1; i < n; i++){
        // for(int j=i; j>0 && arr[j] < arr[j-1]; j--){
        //     swap(arr[j], arr[j-1]);
        // }

        /* 将交换操作改为赋值操作可以提高效率 */

        T key = arr[i];
        int j;
        for (j = i; j > 0 && arr[j-1] > key; j--)
        {
            arr[j] = arr[j - 1];
        }
        arr[j] = key;
    }
}

int main()
{
    int a[10] = {
        10,
        9,
        8,
        4,
        3,
        2,
        17,
        6,
        5,
    };
    insertSort(a, 10);
    for (int i = 0; i < 10; i++)
    {
        cout << a[i] << " ";
    }
    cout << endl;
}